Add citation context extraction from PDF Related Work sections

Muskan244 opened the pull request:
Closes #14085
<!-- LINK THE ISSUE WITH THE "Closes" KEYWORD. Example: Closes https://github.com/JabRef/jabref/issues/13109 OR Closes #13109 -->

<!-- In about one to three sentences, describe the changes you have made: what, where, why, ... Summarize changes and DO NOT list modified classes one-by-one. (REPLACE THIS LINE) -->
This PR implements citation context extraction from PDF Related Work sections. When viewing an entry with an attached PDF, users can extract citations and their surrounding context, match them to library entries, and automatically add the context descriptions to cited entries' comment fields in the format `[sourceCitationKey]: context description`.
Changes:
  - Add new "Show tab 'Citation contexts'" setting in Entry editor preferences
  - Create CitationContextIntegrationService to orchestrate PDF extraction and matching
  - Create PdfSectionExtractor to identify Related Work and References sections
  - Create CitationContextExtractor to parse citation markers and surrounding text
  - Add UI component to display extraction results with clickable apply functionality
  - Contexts are written to cited entries' comment-{username} field when applied

<!-- NOTE: If your work is not yet complete, please open a draft pull request. In that case, outline your intended next steps. Do you need feedback? Will you continue in parallel? ... -->

### Steps to test

<!-- Describe how reviewers can test this fix/feature. Ideally, think of how you would guide a beginner user of JabRef to try out your change. -->
<!-- You can add screenshots or videos (using Loom - https://www.loom.com or by just adding .mp4 files). -->
<!-- (REPLACE THIS PARAGRAPH) -->
1. Open JabRef and load a library with at least one entry that has a PDF attached
2. Go to **Options → Preferences → Entry editor** and ensure "Show tab 'Citation contexts'" is enabled
3. Open an entry that has an academic PDF attached (ideally one with a Related Work or Literature Review section)
4. Click on the **Citation contexts** tab in the entry editor
5. Click **"Extract from this PDF"** button
6. Wait for extraction to complete - you should see a table with:
     - Citation markers found (e.g., "(Smith 2020)", "[1]")
     - Cited entry (matched library entry or "Not found")
     - Context text (the surrounding sentences)
     - Status (Existing, New entry, or Unmatched)
     <img width="1470" height="838" alt="Screenshot 2025-12-14 at 6 27 01 PM" src="https://github.com/user-attachments/assets/2342c62e-c861-441d-a048-1b5d714c05e3" />
7. Select the contexts you want to apply using the checkboxes
8. Click **"Apply selected"** button
9. Check the cited entries in your library - they should now have a `comment-{yourusername}` field containing the context in format: `[SourcePdfKey]: description text`
<img width="1467" height="836" alt="Screenshot 2025-12-14 at 6 27 33 PM" src="https://github.com/user-attachments/assets/f688533d-eaec-4d21-ae1d-25db2d6a9c6b" />

<!-- YOU HAVE TO MODIFY THE ABOVE TEXT TO FIT YOUR PR. OTHERWISE, YOUR PR WILL BE CLOSED WITHOUT FURTHER COMMENT. -->

### Mandatory checks

<!--
Go through the checklist below. It is mandatory, even for a draft pull request.

Keep ALL the items. Replace the dots inside [.] and mark them as follows: 
[x] done 
[ ] TODO (yet to be done)
[/] not applicable
-->

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] I manually tested my changes in running JabRef (always required)
- [x] I added JUnit tests for changes (if applicable)
- [x] I added screenshots in the PR description (if change is visible to the user)
- [x] I described the change in `CHANGELOG.md` in a way that is understandable for the average user (if change is visible to the user)
- [/] I checked the [user documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request updating file(s) in <https://github.com/JabRef/user-documentation/tree/main/en>.

jabref-machine commented:
Your pull request modified git submodules.

Please follow our [FAQ on submodules](https://devdocs.jabref.org/code-howtos/faq.html#submodules) to fix.


<!-- ghprcomment
Check PR Modifications
Submodules not modified
-->

calixtus commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/CitationCommentWriter.java:
```diff
@@ -0,0 +1,251 @@
+        }
+
+        return (double) matchCount / Math.max(words1.length, words2.length);
+    }
```
org.jabref.logic.util.strings.StringSimilarity

calixtus commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/CitationMatcher.java:
```diff
@@ -0,0 +1,402 @@
+    private String normalizeMarker(String marker) {
+        return marker
+                .replaceAll("[\\[\\](){}]", "")
+                .replaceAll("\\s+", " ")
```
use constants

calixtus commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/LibraryEntryResolver.java:
```diff
@@ -0,0 +1,431 @@
+        }
+
+        return cleaned;
+    }
```
Duplication

calixtus commented on jabgui/src/main/java/org/jabref/gui/preferences/JabRefGuiPreferences.java:
```diff
@@ -424,6 +429,7 @@ public EntryEditorPreferences getEntryEditorPreferences() {
                 getBoolean(AUTOLINK_FILES_ENABLED),
                 EntryEditorPreferences.JournalPopupEnabled.fromString(get(JOURNAL_POPUP)),
                 getBoolean(SHOW_SCITE_TAB),
+                getBoolean(SHOW_CITATION_CONTEXT_TAB),
```
There already have been made changes here, so you will need to fix merge conflicts

Muskan244 commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/CitationContextExtractor.java:
```diff
@@ -0,0 +1,131 @@
+
+    private static final Pattern AUTHOR_KEY_PATTERN = Pattern.compile(
+            "\\[([A-Z][a-zA-Z]+\\d{2,4}[a-z]?)\\]"
+    );
```
You're right, the regex-based approach is too limited. I'll refactor `CitationContextExtractor` to use the existing AI infrastructure instead of regex patterns

Muskan244 commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/CitationMatcher.java:
```diff
@@ -0,0 +1,402 @@
+        }
+
+        return parts[parts.length - 1].replaceAll("[^a-zA-Z]", "").toLowerCase(Locale.ROOT);
+    }
```
Now using `AuthorListParser` to properly parse author names and extract the first author's family name via `getFamilyName()`. Added a fallback to simple string  extraction in case parsing fails.

jabref-machine commented:
Your code currently does not meet [JabRef's code guidelines](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). We use [Checkstyle](https://checkstyle.sourceforge.io/) to identify issues. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Checkstyle (pull_request)" and click on it.

In case of issues with the import order, double check that you [activated Auto Import](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-proper-import-cleanup). You can trigger fixing imports by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to trigger [Optimize Imports](https://www.jetbrains.com/guide/tips/optimize-imports/).

Please carefully follow [the setup guide for the codestyle](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). Afterwards, please [run checkstyle locally](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#run-checkstyle) and fix the issues, commit, and push.


<!-- ghprcomment
Source Code Tests
Checkstyle
-->

calixtus commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/LibraryEntryResolver.java:
```diff
@@ -0,0 +1,431 @@
+
+        java.util.regex.Pattern authorYearPattern = java.util.regex.Pattern.compile(
+                "([A-Za-z'\\-]+)(?:\\s+(?:et\\s+al\\.?|and|&)\\s+[A-Za-z'\\-]+)*\\s*(\\d{4})[a-z]?"
+        );
```
Why all the qualifiers?

koppor commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/CitationCommentWriter.java:
```diff
@@ -0,0 +1,229 @@
+        this.username = null;
+    }
+
+    public CitationCommentWriter(String username) {
```
Please use preferences. Look around for `getUserAndHost`. Maybe, you need to adapt something there.

koppor reviewed (changes requested):
Currently lacking resources for a full review (you are seeing 50+ open PRs)

Just a quick scroll through.

InAnYan commented on CHANGELOG.md:
```diff
@@ -32,6 +32,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 - We added the option to change the Git username and PAT in Network Preferences. [#14509](https://github.com/JabRef/jabref/pull/14509)
 - When parsing a plain text citation, we added support for recognizing and extracting arXiv identifiers. [#14455](https://github.com/JabRef/jabref/pull/14455)
 - We introduced a new "Search Engine URL Template" setting in Preferences to allow users to customize their search engine URL templates [#12268](https://github.com/JabRef/jabref/issues/12268)
+- We added a new "Citation contexts" tab in the entry editor that extracts citation contexts from attached PDFs and adds them to cited entries' comment fields. [#14085](https://github.com/JabRef/jabref/issues/14085)
```
```suggestion
- We added a new "Citation contexts" tab in the entry editor that extracts citation contexts from attached PDFs. [#14085](https://github.com/JabRef/jabref/issues/14085)
```

- I think able to view them is the main feature
- Easier and faster to read, everything else are details

InAnYan commented on jablib/src/main/java/org/jabref/logic/ai/AiPreferences.java:
```diff
@@ -125,16 +125,18 @@ public AiPreferences(boolean enableAi,
         this.apiKeyChangeListener = () -> {
         };
 
-        this.templates = Map.of(
```
If possible, please revert to old style

koppor commented:
@InAnYan I try to write down the issue differently.

As researcher, one does different things. When diving into a topic, one reads papers. Really reading. In depth-reading. Maybe using https://sioyek.info/. Really spending much time...

A paper consists briefly of title, abstract, text, references. One can use AI to summarize. But in human-written papers, one has the related work section. The smart authors summarized the referenced paper. With human-readable text. Sometimes, this text is very interesting. Sometimes, one wants to add the referenced paper to the library.

More mathematical:

related work: set of (summary, reference).

The issue is about copying summary and reference into JabRef.

Note: Not all references in the related work section are interesting. The first idea was that I as human reader select text + reference - and the tool imports. Thereby, I forgot that JabRef does not know about the reference itself when pasting using the new entry dialog. Then, the solution at hand is nice. -- Handling of plain text is future work.

Muskan244 commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/CitationCommentWriter.java:
```diff
@@ -0,0 +1,229 @@
+
+    public CitationCommentWriter() {
+        this.commentField = StandardField.COMMENT;
+        this.username = null;
```
Removed the no-argument constructor. Tests were updated to use a constant test username.
Also replaced `Objects.requireNonNull` checks with `@NonNull` annotations across the class and simplified 	`getUsername()` to return a `String` directly instead of an `Optional` as the username is always present.

palukku commented on jablib/src/main/resources/l10n/JabRef_en.properties:
```diff
@@ -3053,6 +3053,47 @@ Metrics\ provided\ by\ scite.ai=Metrics provided by scite.ai
+Found\ %0\ citation\ context(s),\ but\ none\ could\ be\ matched\ to\ library\ entries.\ Ensure\ the\ cited\ papers\ are\ in\ your\ library\ with\ matching\ author\ names\ and\ years.=Found %0 citation context(s), but none could be matched to library entries. Ensure the cited papers are in your library with matching author names and years.
+Found\ %0\ citation\ context(s)...=Found %0 citation context(s)...
+Found\ %0\ citation\ context(s)\:\ %1\ matched,\ %2\ unmatched.\ Select\ which\ to\ apply.=Found %0 citation context(s): %1 matched, %2 unmatched. Select which to apply.
+New\ entry=New entry
```
Is duplicate, we already have "New\ Entry"

palukku commented on jablib/src/main/java/org/jabref/model/pdf/PdfDocumentSections.java:
```diff
@@ -0,0 +1,55 @@
+            "related studies",
+            "theoretical background",
+            "prior work"
+    );
```
Maybe this could be configurable so I can use it in other languages as well (could be a follow up pr, thats fine)

palukku commented on jablib/src/main/java/org/jabref/model/citation/ReferenceEntry.java:
```diff
@@ -0,0 +1,263 @@
+        Objects.requireNonNull(volume, "Volume optional cannot be null");
+        Objects.requireNonNull(pages, "Pages optional cannot be null");
+        Objects.requireNonNull(doi, "DOI optional cannot be null");
+        Objects.requireNonNull(url, "URL optional cannot be null");
```
We use jspecify for nullness checks: https://devdocs.jabref.org/decisions/0052-jspecify-nullable-annotations.html

palukku commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/PdfReferenceParser.java:
```diff
@@ -0,0 +1,521 @@
+            case AUTHOR_YEAR ->
+                    references.addAll(splitByBlankLinesOrIndentation(normalizedText));
+            case AUTHOR_KEY ->
+                    references.addAll(splitByPattern(normalizedText, Pattern.compile("(?=\\[[A-Z][a-zA-Z]+\\d{2,4}[a-z]?\\])")));
```
Can't you extract those too or why can't you use already existing patterns like `AUTHOR_KEY_MARKER_PATTERN`?

calixtus commented on jabgui/src/main/java/org/jabref/gui/entryeditor/citationcontexttab/CitationContextComponent.java:
```diff
@@ -0,0 +1,508 @@
+                                      allMatched.addAll(matched);
+
+                                      int currentSize = allMatched.size();
+                                      Platform.runLater(() -> {
```
UiTaskExecutor

calixtus commented on jabgui/src/main/java/org/jabref/gui/entryeditor/citationcontexttab/CitationContextComponent.java:
```diff
@@ -0,0 +1,508 @@
+        private final SimpleStringProperty status;
+        private final CitationContextIntegrationService.MatchedContext matchedContext;
+
+        public ExtractedContextRow(String citationMarker, String targetEntry, String contextText,
```
one arg per line

calixtus commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/CitationCommentWriter.java:
```diff
@@ -0,0 +1,208 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CitationCommentWriter {
```
You can also use NullMarked

calixtus commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/CitationContextExtractor.java:
```diff
@@ -0,0 +1,131 @@
+    private List<SentencePosition> splitIntoSentences(String text) {
+        List<SentencePosition> sentences = new ArrayList<>();
+
+        Pattern sentenceEnd = Pattern.compile(
```
I imagine this will be run for every time a doc is opened? Maybe put that into the constructor?

Muskan244 commented on jablib/src/main/java/org/jabref/logic/citation/contextextractor/CitationContextIntegrationService.java:
```diff
@@ -0,0 +1,203 @@
+            textToAnalyze = sections.fullText();
+        }
+
+        if (llmContextExtractor != null) {
```
Also refactored to fail fast principle.

jabref-machine commented:
Your code currently does not meet [JabRef's code guidelines](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). We use [Checkstyle](https://checkstyle.sourceforge.io/) to identify issues. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Checkstyle (pull_request)" and click on it.

In case of issues with the import order, double check that you [activated Auto Import](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-proper-import-cleanup). You can trigger fixing imports by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to trigger [Optimize Imports](https://www.jetbrains.com/guide/tips/optimize-imports/).

Please carefully follow [the setup guide for the codestyle](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). Afterwards, please [run checkstyle locally](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#run-checkstyle) and fix the issues, commit, and push.


<!-- ghprcomment
Source Code Tests
Checkstyle
-->

koppor commented on CHANGELOG.md:
```diff
@@ -15,6 +15,8 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 - Improved merge dialog decisions for the `month` field by using a new plausibility comparator. JabRef now prefers normalized months (e.g., `#jun#`) or integers over unnormalized strings. [#14457](https://github.com/JabRef/jabref/issues/14457)
 - We added web search capability to ScienceDirect using the [Scopus](https://www.scopus.com/) Search API, enabling users to search for academic literature directly from JabRef. [#14421](https://github.com/JabRef/jabref/issues/14421)
 - We fixed an issue where a redundant validation listener was causing duplicate error dialogs when invalid BibTeX source was detected in the SourceTab. [#14805](https://github.com/JabRef/jabref/issues/14805)
+- We added support for selecting citation fetcher in Citations Tab. [#14430](https://github.com/JabRef/jabref/issues/14430)
```
Double entry. Remove.

koppor replied:
Is this AI generated text??? Really, don't play games with us.

The GitHub comments are always on the line before. Thus, `14430` is meant.

Look three lines down - this is not too hard to see

<img width="793" height="197" alt="Image" src="https://github.com/user-attachments/assets/b52d18e6-5d8f-4a9e-ba6a-108301b25ae9" />

Muskan244 commented:
> Here and there, it feels like AI generated. `NullMarked` / `Nullable` not used, but == null.
> 
> Magic numbers instead of constants.
> 
> But a good start, but a senior needs to go through.

Thanks for the feedback, Oliver! You are right, I have addressed both issues: Added `@NullMarked` annotation to all citation context classes and used `@Nullable` for variables that can be null instead of raw null checks, and extracted all  magic numbers into named constants with descriptive names.

calixtus commented:
Stop using ai to communicate with us or you get yourself banned. We don't need a human communication layer for an LLM.
Speak directly, speak what you think. Even if it's not the best english.
